### PR TITLE
Lpal 677 DNS Resolver Firewall, and Internal Private DNS Namespace changes for Service Discovery

### PR DESCRIPTION
**Pull request by williamfalconeruk**
## Purpose

To support the requirement to monitor Egress out of AWS, and to create alert records for cloudwatch logs.
Additionally - support in dev the use of `.${ACCOUNT_NAME}.internal` domain name. See ADR 0009 in this PR.

Fixes LPAL-677

## Approach

- Add DNS firewall egress monitoring Terraform
- Update Service Discovery for API to use `.${ACCOUNT_NAME}.internal` suffix as this is more canonical and is easer to filter in the Firewall.
- ADR to cover the proposed change.
There will be a further ticket to implement pre and production changes, which will be done within agreed downtime.
## Learning

See https://github.com/ministryofjustice/opg-use-an-lpa/pull/1436 and 
https://github.com/ministryofjustice/opg-use-an-lpa/pull/1382

## Checklist

* [X] I have performed a self-review of my own code
* [X] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [ ] I have added tests to prove my work
* [X] I have added mandatory tags to terraformed resources, where possible
* [ ] If I added a package.json or composer.json, I also made sure this is included in the script in `.github/workflows/dependabot-update.yml`
* [ ] The product team have tested these changes

### CODE REVIEW

File: docs/architecture/0009-change-internal-namespace.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+Date: 31/03/2022
+
+## Status
+
+Proposed (31/03/2022)
+
+## Context
+
+Currently it is hard to filter the `-internal` suffix for service discovery namespaces in Route 53 Resolver DNS firewall. the Domain name used  does not have it's own "parts", and presents issues when you have multiple ephemeral environments in the AWS dev account, when setting up rules.
```
Comment (johnnolan) [inline docs/architecture/0009-change-internal-namespace.md:11]:
```suggestion
Currently it is hard to filter the `-internal` suffix for service discovery namespaces in Route 53 Resolver DNS firewall. the Domain name used does not have it's own "parts", and presents issues when you have multiple ephemeral environments in the AWS dev account, when setting up rules.
```

File: docs/architecture/0009-change-internal-namespace.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+For details of the route 53 Reolver DNS Firewall see: <https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-dns-firewall.html>
```
Comment (johnnolan) [inline docs/architecture/0009-change-internal-namespace.md:13]:
```suggestion
For details of the route 53 Reolver DNS Firewall see: [https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-dns-firewall.html](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-dns-firewall.html)

```

File: docs/architecture/0009-change-internal-namespace.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+For details of the service discovery private namespaces see:
+<https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-discovery.html>
```
Comment (johnnolan) [inline docs/architecture/0009-change-internal-namespace.md:15]:
```suggestion
For details of the service discovery private namespaces see: [https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-discovery.html](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/service-discovery.html)
```

File: docs/architecture/0009-change-internal-namespace.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+## Proposal
+
+In order to do proper filtering on the DNS Firewall, we need to change the DNS entries used for Service discovery, so that it is easier to support.
+
+We propose that this is changed from the current format of
+
+``` HCL
+
+api.${environment_name}-internal
+```
```
Comment (johnnolan) [inline docs/architecture/0009-change-internal-namespace.md:26]:
```suggestion
``` HCL
api.${environment_name}-internal
```
```

File: docs/architecture/0009-change-internal-namespace.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+To
+
+``` HCL
+api.${environment_name}.${account_name}.internal  # development only
+api.${account_name}.internal # production and preproduction only
+```
+
+Where:
+
+- `${environment_name}` refers to ephemeral  as defined by terraform e.g. `PR1234`
```
Comment (johnnolan) [inline docs/architecture/0009-change-internal-namespace.md:37]:
```suggestion
- `${environment_name}` refers to ephemeral as defined by terraform e.g. `PR1234`
```

**Review by johnnolan — CHANGES_REQUESTED** [review]
This looks really good and sensible. Would it be worth also adding a shorterened version with a link to this in the https://docs.opg.service.justice.gov.uk/documentation/adrs/adr-002.html#adr-002-application-domain-names page?

**Review by johnnolan — APPROVED** [review]

**Review by johnnolan — DISMISSED** [review]

**Review by Raffers — DISMISSED** [review]

**Review by johnnolan — APPROVED** [review]

**Review by Raffers — APPROVED** [review]

**Pull request merged**